pubsub-mapped-api/src/main/java/com/google/pubsub/clients/consumer/ack/MessageDispatcher.java

```diff
@@ -542,22 +535,20 @@ private void setupNextAckDeadlineExtensionAlarm(Instant expiration) {
     }
   }
 
+  //TODO: do we really need the sync?
   void acknowledgePendingMessages(boolean sync, Long offset) {
-    Map<String, AckHandler> acksToSend = new HashMap<>();
+    Map<String, AckHandler> acksToSend;
     synchronized (pendingAcks) {
-      if (!pendingAcks.isEmpty()) {
-        for (Entry<String, AckHandler> pair : pendingAcks.entrySet()){
-          acksToSend.put(pair.getKey(), pair.getValue());
-        }
-        logger.log(Level.FINER, "Sending {0} acks", acksToSend.size());
-      }
+      acksToSend = pendingAcks;
+      pendingAcks = new ConcurrentHashMap<>();
     }
+    logger.log(Level.FINER, "Sending {0} acks", acksToSend.size());
 
-    if(offset != null) {
+    if (offset != null) {
```

Comment:
this will not work. after you call filterAcksBeforeOffset all messages that are not before this offset will never be acked and their deadlines will be still extended. in this case you would have to put all removed values back to pendingAcks (as it is empty now).

Reply:
Fixed, check it again now.